### PR TITLE
Fix deadlock in FunctionRef, #26326

### PR DESCRIPTION
**Pull request by patriknw**
Refs #26326

See thread dump in issue.

Don't squash when merging since test reproducer was from contributor.

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/ActorRef.scala [hunk 1]
```diff
@@ -732,83 +732,106 @@ private[akka] final class FunctionRef(
...
-      case OptionVal.Some(watchedBy) ⇒
-        if (watchedBy.nonEmpty) {
-          watchedBy foreach sendTerminated(ifLocal = false)
-          watchedBy foreach sendTerminated(ifLocal = true)
-        }
+    val toUnwatch = this.synchronized {
+      _watchedBy match {
+        case OptionVal.Some(watchedBy) ⇒
+          if (watchedBy.nonEmpty) {
+            watchedBy foreach sendTerminated(ifLocal = false)
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/actor/ActorRef.scala]: Are these two still ok to be inside the synchronized block?

Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/ActorRef.scala]: good catch, my bad

Comment (patriknw) [reply akka-actor/src/main/scala/akka/actor/ActorRef.scala]: fixed in 4310184

**Review by johanandren — APPROVED** [review]

**Review by chbatey — APPROVED** [review]
LGTM

**Pull request merged**